File: src/main/java/ltdjms/discord/discord/domain/EmbedView.java
package ltdjms.discord.discord.domain;

import net.dv8tion.jda.api.entities.MessageEmbed;

import java.awt.Color;
import java.util.List;

/**
 * Embed 視圖的不可變資料結構
 *
 * <p>此 record 用於傳遞 Embed 的資料，提供與 JDA MessageEmbed 的解耦。
 *
 * <p>長度限制驗證由 {@link DiscordEmbedBuilder#buildPaginated(EmbedView)} 負責，
 * 以支援自動分頁功能。
 *
 * @param title 標題
 * @param description 描述
 * @param color 顏色
 * @param fields 欄位列表
 * @param footer Footer 文字
 */
public record EmbedView(
        String title,
        String description,
        Color color,
        List<FieldView> fields,
        String footer
) {
    /**
     * 欄位視圖
     *
     * @param name 名稱
     * @param value 值
     * @param inline 是否內聯顯示
     */
    public record FieldView(
            String name,
            String value,
            boolean inline
    ) {

        /**
         * 轉換為 JDA MessageEmbed.Field
         *
         * @return MessageEmbed.Field 物件
         */
        public MessageEmbed.Field toJdaField() {
            return new MessageEmbed.Field(name, value, inline);
        }
    }
}

File: packages/shared/src/discord/domain/embed-view.ts
import { EmbedBuilder, ButtonBuilder } from 'discord.js';

/**
 * Immutable data structure for Embed view data.
 * Provides decoupling from discord.js MessageEmbed.
 * Matches Java EmbedView record.
 */
export interface EmbedView {
  readonly title?: string;
  readonly description?: string;
  readonly color?: number;
  readonly fields?: FieldView[];
  readonly footer?: string;
}

/**
 * Field view for an embed field.
 * Matches Java EmbedView.FieldView record.
 */
export interface FieldView {
  readonly name: string;
  readonly value: string;
  readonly inline: boolean;
}

/**
 * Immutable data structure for Button view data.
 * Matches Java ButtonView record.
 */
export interface ButtonView {
  readonly id: string;
  readonly label: string;
  readonly style: ButtonStyle;
  readonly disabled: boolean;
}

/** Button style enum matching discord.js ButtonStyle. */
export enum ButtonStyle {
  PRIMARY = 1,
  SECONDARY = 2,
  SUCCESS = 3,
  DANGER = 4,
  LINK = 5,
}

/**
 * Creates a ButtonView with length validation on id and label.
 * Discord limits: id max 100 chars, label max 80 chars.
 * Throws if either exceeds the limit.
 */
export function createButtonView(
  id: string,
  label: string,
  style: ButtonStyle = ButtonStyle.PRIMARY,
  disabled = false,
): ButtonView {
  if (id.length > 100) {
    throw new Error(`ButtonView id exceeds 100 character limit: ${id.length} chars`);
  }
  if (label.length > 80) {
    throw new Error(`ButtonView label exceeds 80 character limit: ${label.length} chars`);
  }
  return { id, label, style, disabled } as ButtonView;
}

/**
 * Converts an EmbedView to a discord.js EmbedBuilder.
 * Pure data transformation — no side effects.
 * Matches spec R8.5 requirement for paginated embed construction support.
 */
export function toDiscordJsEmbed(view: EmbedView): EmbedBuilder {
  const embed = new EmbedBuilder();
  if (view.title) embed.setTitle(view.title);
  if (view.description) embed.setDescription(view.description);
  if (view.color !== undefined) embed.setColor(view.color);
  if (view.footer) embed.setFooter({ text: view.footer });
  if (view.fields) {
    embed.addFields(
      view.fields.map((f) => ({ name: f.name, value: f.value, inline: f.inline })),
    );
  }
  return embed;
}

/**
 * Converts a ButtonView to a discord.js ButtonBuilder.
 * Pure data transformation — no side effects.
 * Matches spec R8.5 requirement for paginated embed construction support.
 */
export function toDiscordJsButton(view: ButtonView): ButtonBuilder {
  return new ButtonBuilder()
    .setCustomId(view.id)
    .setLabel(view.label)
    .setStyle(view.style as unknown as number)
    .setDisabled(view.disabled);
}
